File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro/DarkWilderness.java
package com.asiantech.intern20summer1.intro;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class DarkWilderness {
    /**
     * Write a funtion that returns the number of days that it will take
     * for the plant to reach / pass desiredHeight.
     */
    int growingPlant(int upSpeed, int downSpeed, int desiredHeight) {
        int height = 0;
        int day = 0;

        while (true) {
            height += upSpeed;
            day++;
            if (height >= desiredHeight) {
                break;
            } else {
                height -= downSpeed;
            }
        }

        return day;
    }

    /**
     * Write a funtion that returns the total maximum value of the items you can take.
     */
    int knapsackLight(int value1, int weight1, int value2, int weight2, int maxW) {
        int result = 0;
        int[] values = new int[]{value1, value2};
        Map map = new HashMap();

        Arrays.sort(values);
        map.put(value1, weight1);
        map.put(value2, weight2);

        for (int i = values.length - 1; i >= 0; i--) {
            if (maxW >= (int) map.get((Object) values[i])) {
                result += values[i];
                maxW -= (int) map.get((Object) values[i]);
            }
        }

        return result;
    }

    /**
     * Write a funtion that returns the longest prefix which contains only digits.
     */
    String longestDigitsPrefix(String inputString) {
        String result = "";
        for (int i = 0; i < inputString.length(); i++) {
            if (Character.isDigit(inputString.charAt(i))) {
                result += inputString.charAt(i);
            } else {
                break;
            }
        }

        return result;
    }

    /**
     * Write a funtion that returns the sum of digits.
     */
    int sumOfDigits(int n) {
        int sum = 0;
        while (n > 0) {
            sum += (n % 10);
            n /= 10;
        }

        return sum;
    }

    /**
     * Write a funtion that returns its digit degree.
     */
    int digitDegree(int n) {
        int count = 0;
        while (n >= 10) {
            count++;
            n = sumOfDigits(n);
        }

        return count;
    }

    /**
     * Write a funtion that checks whether the bishop can capture the pawn in one move.
     */
    boolean bishopAndPawn(String bishop, String pawn) {
        int i1Bishop = 7 - (bishop.charAt(0) - 97);
        int i2Bishop = bishop.charAt(1) - 49;
        int i1Pawn = 7 - (pawn.charAt(0) - 97);
        int i2Pawn = pawn.charAt(1) - 49;

        if (i1Pawn > i1Bishop && i2Pawn > i2Bishop) {
            if ((i1Pawn - i1Bishop) == (i2Pawn - i2Bishop)) {
                return true;
            }
        } else if (i1Pawn < i1Bishop && i2Pawn < i2Bishop) {
            if ((i1Bishop - i1Pawn) == (i2Bishop - i2Pawn)) {
                return true;
            }
        } else if (i1Pawn > i1Bishop && i2Pawn < i2Bishop) {
            if ((i1Pawn - i1Bishop) == (i2Bishop - i2Pawn)) {
                return true;
            }
        } else if (i1Pawn < i1Bishop && i2Pawn > i2Bishop) {
            if ((i1Bishop - i1Pawn) == (i2Pawn - i2Bishop)) {
                return true;
            }
        }

        return false;
    }
}

File: app/src/at-phuongle/java/com/asiantech/intern20summer1/kotlin/DarkWilderness.kt
package com.asiantech.intern20summer1.kotlin

object DarkWilderness {
    private const val TEN = 10
    private const val SEVEN = 7
    private const val a = 97
    private const val TO_NUMBER = 49

    fun growingPlant(upSpeed: Int, downSpeed: Int, desiredHeight: Int): Int {
        var height: Int = 0
        var day: Int = 0

        while (true) {
            height += upSpeed
            day++
            if (height >= desiredHeight) {
                break
            } else {
                height -= downSpeed
            }
        }

        return day
    }

    fun knapsackLight(value1: Int, weight1: Int, value2: Int, weight2: Int, maxW: Int): Int {
        var maxWeight: Int = maxW
        var result: Int = 0
        var values: Array<Int> = arrayOf(value1, value2)
        var map: MutableMap<Int, Int> = hashMapOf()

        values.sort()
        map[value1] = weight1
        map[value2] = weight2

        for (i in (values.size - 1) downTo 0) {
            if (maxWeight >= map.getValue(values[i])) {
                result += values[i]
                maxWeight -= map.getValue(values[i])
            }
        }

        return result
    }

    fun longestDigitsPrefix(inputString: String): String {
        var result: String = ""
        for (i in inputString.indices) {
            if (inputString[i].isDigit()) {
                result += inputString[i]
            } else {
                break
            }
        }

        return result
    }

    fun sumOfDigits(n: Int): Int {
        var number: Int = n
        var sum: Int = 0

        while (number > 0) {
            sum += (number % TEN)
            number /= TEN
        }

        return sum
    }

    fun digitDegree(n: Int): Int {
        var number: Int = n
        var count: Int = 0

        while (number >= TEN) {
            count++
            number = sumOfDigits(number)
        }

        return count
    }

    fun bishopAndPawn(bishop: String, pawn: String): Boolean {
        var check: Boolean = false
        var i1Bishop: Int = SEVEN - (bishop[0].toInt() - a)
        var i2Bishop: Int = bishop[1].toInt() - TO_NUMBER
        var i1Pawn: Int = SEVEN - (pawn[0].toInt() - a)
        var i2Pawn: Int = pawn[1].toInt() - TO_NUMBER

        if (i1Pawn > i1Bishop && i2Pawn > i2Bishop) {
            if ((i1Pawn - i1Bishop) == (i2Pawn - i2Bishop)) {
                check = true
            }
        } else if (i1Pawn < i1Bishop && i2Pawn < i2Bishop) {
            if ((i1Bishop - i1Pawn) == (i2Bishop - i2Pawn)) {
                check = true
            }
        } else if (i1Pawn > i1Bishop && i2Pawn < i2Bishop) {
            if ((i1Pawn - i1Bishop) == (i2Bishop - i2Pawn)) {
                check = true
            }
        } else if (i1Pawn < i1Bishop && i2Pawn > i2Bishop) {
            if ((i1Bishop - i1Pawn) == (i2Pawn - i2Bishop)) {
                check = true
            }
        }

        return check
    }
}
